src/core/http1/util.rs

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (C) 2023-2024 Fastly, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use crate::buffer::{Buffer, VecRingBuffer};
+use crate::future::{AsyncRead, AsyncReadExt};
+use std::cmp;
+use std::future::Future;
+use std::io;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+pub const HEADERS_MAX: usize = 64;
+
+// return the capacity increase
+pub fn resize_write_buffer_if_full<F>(
+    buf: &mut VecRingBuffer,
+    block_size: usize,
+    blocks_max: usize,
+    reserve: F,
+) -> usize
+where
+    F: Fn() -> bool,
+{
+    assert!(blocks_max >= 2);
+
+    // all but one block can be used for writing
+    let allowed = blocks_max - 1;
+
+    if buf.remaining_capacity() == 0 && buf.capacity() < block_size * allowed && reserve() {
```

Comment:
this assumes block_size * allowed will not overflow. we can use `checked_mul` to ensure